### ISSUE TITLE
can't pass trial_end timestamp to Stripe

### ISSUE THREAD

**Issue by enkay** [laravel/cashier#345]
I have a subscription that doesn't require a card upfront. As per the documentation, I set the trial_ends_at on the user model when they start their trial.

Now if the user upgrades before the end of the trial, I don't want Stripe to charge them until the end of the trial, so I want to set the **trial_end** parameter on the request.

Right now it seems the only way to set **trial_end**  is through the **trialDays()** method which is not very convenient as I already have the trial end timestamp, and I don't want cashier to calculate a new date based on the current time.

It seems there was a hidden method **trialFor()**  that could do this in previous versions but it was removed  and doesn't work when I try it.

**Comment by kyranb** [comment]
Agree, this behaviour seems really off and has caused a lot of confusion from what I've seen on GitHub, The Forums, Reddit etc.

**Comment by enkay** [comment]
I'm surprised this hasn't been brought up before as it seems like a very common subscription model for saas businesses: offer a free trial, and try to get users to upgrade to a paid subscription before their trial ends.

Is there any plans to better support this model? Or is there another workaround?

**Comment by themsaid** [comment]
@enkay  can you please share some code samples? How do you instruct cashier to upgrade the user to a paid plan??

**Comment by joshgriggs** [comment]
I have now also run into this problem. The context is as follows:

- A user signs up to our application without card information up front. The registration code manually creates their trial end by updating the trial_ends_at column in the database for that user. 
- When the user visits the billing page, they can see that they are on trial and can choose to add their card information to ensure that their subscription remains live after their trial end date
- When they submit their card information, the subscription is then created with the trialDays() method, with the number of days currently being calculated as the difference in days from their trial end date from now. This is problematic, because if the trial ends at date is 9 days 23 hours and 59 minutes away, it is calculated as being 9 days away thus effectively reducing the users' trial by a day.

It would be great if the trialFor() method could be re-implemented so that a timestamp / carbon date could be used instead of specifying days.

**Comment by codegeek1001** [comment]
There is a skipTrial() function that lets you end a trial immediately which essentially nulls the trial_ends field.

**Comment by enkay** [comment]
@codegeek1001 I don't want to cancel the trial, I want to start charging once their trial is over. I need a method to set the underlying Stripe parameter **trial_end** manually.

**Comment by codegeek1001** [comment]
I understand what you are saying but in practice, when a client "upgrades", they are charged right away and it does not care how many trial days still remain. In other words, they become a paying client effective immediately. 

Not trying to say that the trialFor method should not be implemented but I have not seen a subscription service where upgrading the plan does not charge my card immediately.

**Comment by kyranb** [comment]
Shopify or Intercom are both  examples that do this. You can add a payment
method and upgrade, but your card is not charged until the original trial
is ended.

**Comment by enkay** [comment]
@codegeek1001 As pointed out by @kyranb that is simply not true, a lot of services do this. Also as far as business practices go, it also encourages customers to sign up early instead of waiting until the last day of their trial.

**Comment by enkay** [comment]
Any news on this?

TLDR: There is currently no way to manually set the Stripe parameter **trial_end** with Cashier. It's a problem because it makes it impossible to support [this](https://github.com/laravel/cashier/issues/345#issuecomment-268499157) very common SAAS free trial / upgrade model.

**Comment by mrsimonbennett** [comment]
@enkay A method has been merged in for this now. 

```
$user->newSubscription($planName, $planName)->trailUntil($user->trial_ends_at)->create(
    $request->get('stripeToken')
);
```

**Comment by joshgriggs** [comment]
@mrsimonbennett Brilliant, thanks!
